src/wolfio.c

```diff
@@ -2714,4 +2714,382 @@ int wolfSSL_SetIO_LwIP(WOLFSSL* ssl, void* pcb,
+            /* Can't send whilst we are receiving */
+        if (isotp_ctx->state != ISOTP_CONN_STATE_IDLE) {
+            return WOLFSSL_ERROR_WANT_READ;
+        }
+        isotp_ctx->state = ISOTP_CONN_STATE_RECEIVING;
+        do {
+            ret = isotp_ctx->recv_fn(&isotp_ctx->frame, isotp_ctx->arg,
+                    ISOTP_DEFAULT_TIMEOUT);
+        } while (ret == 0);
+        if (ret == 0) {
+            isotp_ctx->state = ISOTP_CONN_STATE_IDLE;
+            return WOLFSSL_CBIO_ERR_TIMEOUT;
+        } else if (ret < 0) {
+            isotp_ctx->state = ISOTP_CONN_STATE_IDLE;
+            WOLFSSL_MSG("ISO-TP receive error");
+            return WOLFSSL_CBIO_ERR_GENERAL;
+        }
+
+        type = isotp_ctx->frame.data[0] >> 4;
+
+        if (type == ISOTP_FRAME_TYPE_SINGLE) {
+            isotp_ctx->receive_buffer_len =
+                isotp_receive_single_frame(isotp_ctx);
+        } else if (type == ISOTP_FRAME_TYPE_FIRST) {
+            isotp_ctx->receive_buffer_len =
+                isotp_receive_multi_frame(isotp_ctx);
+        } else {
+            /* Should never get here */
+            isotp_ctx->state = ISOTP_CONN_STATE_IDLE;
+            WOLFSSL_MSG("ISO-TP frames out of sequence");
+            return WOLFSSL_CBIO_ERR_GENERAL;
+        }
+        if (isotp_ctx->receive_buffer_len <= 1) {
+            isotp_ctx->state = ISOTP_CONN_STATE_IDLE;
+            return isotp_ctx->receive_buffer_len;
+        } else {
+            isotp_ctx->receive_buffer_ptr = isotp_ctx->receive_buffer;
+        }
+        isotp_ctx->state = ISOTP_CONN_STATE_IDLE;
+    }
+
+    /* Return from the buffer */
+    if (isotp_ctx->receive_buffer_len >= sz) {
+        XMEMCPY(buf, isotp_ctx->receive_buffer_ptr, sz);
+        isotp_ctx->receive_buffer_ptr+= sz;
+        isotp_ctx->receive_buffer_len-= sz;
+        return sz;
+    } else {
+        XMEMCPY(buf, isotp_ctx->receive_buffer_ptr,
+                isotp_ctx->receive_buffer_len);
+        sz = isotp_ctx->receive_buffer_len;
+        isotp_ctx->receive_buffer_len = 0;
+        return sz;
+    }
+}
+
+/* Setup ISPTP
+ *
+ * Parameters:
+ * ssl - The wolfSSL context
+ * ctx - A user created ISOTP context which this function initializes
+ * recv_fn - A user CAN bus receive callback
+ * send_fn - A user CAN bus send callback
+ * delay_fn - A user microsecond granularity delay function
+ * receive_delay - A set amount of microseconds to delay each CAN bus packet
+ * receive_buffer - A user supplied buffer to receive data, recommended that is
+ *                  allocated to ISOTP_DEFAULT_BUFFER_SIZE bytes
+ * receive_buffer_size - The size of receive_buffer
+ * arg - An arbitrary pointer sent to recv_fn and send_fn
+ */
+int wolfSSL_SetIO_ISOTP(WOLFSSL *ssl, isotp_wolfssl_ctx *ctx,
+        can_recv_fn recv_fn, can_send_fn send_fn, can_delay_fn delay_fn,
+        word32 receive_delay, char *receive_buffer, int receive_buffer_size,
+        void *arg)
+{
+    ctx->recv_fn = recv_fn;
```

Comment:
Public API should check the incoming `ctx` is not NULL..

Reply:
That and several other things would cause issues if they are `NULL`, good point.